# **ksproject**

# How many rows of data are in the DataFrame?

In [ ]:
import pandas as pd

ks_db = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/ksprojects.csv")

ks_db.shape

(65418, 13)

There are 65,418 rows and 13 columns in the ks database.

# What are the names and data types of the columns?

In [ ]:
ks_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65418 entries, 0 to 65417
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             65418 non-null  int64  
 1   name           65418 non-null  object 
 2   category       65418 non-null  object 
 3   main_category  65418 non-null  object 
 4   deadline       65418 non-null  object 
 5   goal           65418 non-null  float64
 6   launched       65418 non-null  object 
 7   state          65418 non-null  object 
 8   backers        65418 non-null  int64  
 9   country        65418 non-null  object 
 10  usd_pledged    64673 non-null  float64
 11  currency       65418 non-null  object 
 12  pledged        65418 non-null  float64
dtypes: float64(3), int64(2), object(8)
memory usage: 6.5+ MB


The column names for the ks database are as follows: 'id', 'name', 'category', 'main_category', 'deadline', 'goal', 'launched', 'state', 'backers', 'country', 'usd_pledged', 'currency', and 'pledged'. There are 3 columns containing float data, 2 with integer data, and 8 columns containing objects. 

# Do any of the columns contain null values?

In [ ]:
ks_db.isnull()

id   name  category  ...  usd_pledged  currency  pledged
0      False  False     False  ...        False     False    False
1      False  False     False  ...        False     False    False
2      False  False     False  ...        False     False    False
3      False  False     False  ...        False     False    False
4      False  False     False  ...        False     False    False
...      ...    ...       ...  ...          ...       ...      ...
65413  False  False     False  ...        False     False    False
65414  False  False     False  ...        False     False    False
65415  False  False     False  ...        False     False    False
65416  False  False     False  ...        False     False    False
65417  False  False     False  ...        False     False    False

[65418 rows x 13 columns]

It appears that the ks database contains no null values. 

# Find all successful documentary projects and sort them by the amount pledged. Print the top 10 highest pledges.

In [ ]:
successful_docs = (ks_db['category'] == 'Documentary') & (ks_db['state'] == 'successful')
successful_docs = ks_db.iloc[successful_docs.values]
most_pledged_docs = successful_docs.sort_values(by='pledged', ascending=False)
most_pledged_docs.head(10)

id  ...   pledged
11731  1078982332  ...  384376.0
24166  1162836718  ...  313158.0
12127  1326363492  ...  240022.0
61248  1203470139  ...  186010.0
28003  1205913248  ...  177421.0
27402  1135999291  ...  176746.0
16243  1061853312  ...  167917.0
27268  1228065368  ...  165401.0
43820  1194449879  ...  163930.0
22237   122290911  ...  163891.0

[10 rows x 13 columns]

# Create a new column named average_per_backer and set its value to the total amount pledged divided by the number of backers. What happened to the rows with zero backers? How can this be dealt with?

In [ ]:
ks_db_new_column = ks_db.assign(average_per_backer = ks_db.usd_pledged / ks_db.backers)
ks_db_new_column.head(20)

id  ... average_per_backer
0   1147015301  ...          10.000000
1   1100844465  ...          50.000000
2   1316334968  ...          42.718750
3   1304906577  ...         273.958333
4   1099226462  ...          22.802727
5   1301564845  ...          70.000000
6   1081313636  ...          15.000000
7    112916020  ...          46.250000
8   1130941392  ...           9.333333
9   1125367305  ...          45.350000
10  1143904247  ...          49.434783
11  1163942232  ...          30.000000
12  1167151653  ...          13.666667
13  1170035792  ...          20.000000
14  1178729979  ...          50.000000
15  1074367099  ...           7.000000
16  1291701547  ...          52.745098
17  1001181420  ...         158.892308
18  1018181261  ...                NaN
19  1275895636  ...                NaN

[20 rows x 14 columns]

The rows with 0 backers have NaN values in the 'average_per_backer' column. We can deal with this by assigning 0 for the NaN values in the 'average_per_backer' column. Or we can remove all rows with 0 backers and then create our 'average_per_backer' column.

# Drop all rows with zero backers, then repeat the previous exercise.

In [ ]:
at_least_one_backer = ks_db['backers'] >= 1
ks_db_updated = ks_db.iloc[at_least_one_backer.values]
ks_db_updated = ks_db_updated.assign(average_per_backer = ks_db_updated.usd_pledged / ks_db_updated.backers)
ks_db_updated.head(20)

id  ... average_per_backer
0   1147015301  ...          10.000000
1   1100844465  ...          50.000000
2   1316334968  ...          42.718750
3   1304906577  ...         273.958333
4   1099226462  ...          22.802727
5   1301564845  ...          70.000000
6   1081313636  ...          15.000000
7    112916020  ...          46.250000
8   1130941392  ...           9.333333
9   1125367305  ...          45.350000
10  1143904247  ...          49.434783
11  1163942232  ...          30.000000
12  1167151653  ...          13.666667
13  1170035792  ...          20.000000
14  1178729979  ...          50.000000
15  1074367099  ...           7.000000
16  1291701547  ...          52.745098
17  1001181420  ...         158.892308
20  1024509505  ...         961.000000
21  1317972662  ...          44.736842

[20 rows x 14 columns]

# Create a crosstab to get a count of records for each combination of state and category.

In [ ]:
pd.crosstab(ks_db['state'], ks_db['category'])


category    3D Printing  Academic  Accessories  ...  World Music  Young Adult  Zines
state                                           ...                                 
canceled             17        21           48  ...           31           16      6
failed               59        86          238  ...          174           78     26
live                  3         3           16  ...            2            1      1
successful           43        43          121  ...          168           23     17
suspended             2         4            2  ...            1            0      0
undefined             1         0            1  ...            0            0      0

[6 rows x 158 columns]

# **nycflights13**

# Merge these two DataFrames into a single DataFrame: nycflights13.

In [ ]:
flights = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/flights.csv')
airlines = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/airlines.csv')

nycflights13 = pd.merge(flights, airlines)
nycflights13.info()
nycflights13.head(30)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 336776 entries, 0 to 336775
Data columns (total 20 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   year            336776 non-null  int64  
 1   month           336776 non-null  int64  
 2   day             336776 non-null  int64  
 3   dep_time        328521 non-null  float64
 4   sched_dep_time  336776 non-null  int64  
 5   dep_delay       328521 non-null  float64
 6   arr_time        328063 non-null  float64
 7   sched_arr_time  336776 non-null  int64  
 8   arr_delay       327346 non-null  float64
 9   carrier         336776 non-null  object 
 10  flight          336776 non-null  int64  
 11  tailnum         334264 non-null  object 
 12  origin          336776 non-null  object 
 13  dest            336776 non-null  object 
 14  air_time        327346 non-null  float64
 15  distance        336776 non-null  int64  
 16  hour            336776 non-null  int64  
 17  minute    

year  month  day  ...  minute            time_hour                   name
0   2013      1    1  ...      15  2013-01-01 05:00:00  United Air Lines Inc.
1   2013      1    1  ...      29  2013-01-01 05:00:00  United Air Lines Inc.
2   2013      1    1  ...      58  2013-01-01 05:00:00  United Air Lines Inc.
3   2013      1    1  ...       0  2013-01-01 06:00:00  United Air Lines Inc.
4   2013      1    1  ...       0  2013-01-01 06:00:00  United Air Lines Inc.
5   2013      1    1  ...       0  2013-01-01 06:00:00  United Air Lines Inc.
6   2013      1    1  ...       7  2013-01-01 06:00:00  United Air Lines Inc.
7   2013      1    1  ...       0  2013-01-01 06:00:00  United Air Lines Inc.
8   2013      1    1  ...      27  2013-01-01 06:00:00  United Air Lines Inc.
9   2013      1    1  ...      30  2013-01-01 06:00:00  United Air Lines Inc.
10  2013      1    1  ...      46  2013-01-01 06:00:00  United Air Lines Inc.
11  2013      1    1  ...      36  2013-01-01 06:00:00  United Air Lines Inc.
12  2013      1    1  ...      45  2013-01-01 06:00:00  United Air Lines Inc.
13  2013      1    1  ...      45  2013-01-01 06:00:00  United Air Lines Inc.
14  2013      1    1  ...       0  2013-01-01 07:00:00  United Air Lines Inc.
15  2013      1    1  ...       0  2013-01-01 07:00:00  United Air Lines Inc.
16  2013      1    1  ...       0  2013-01-01 07:00:00  United Air Lines Inc.
17  2013      1    1  ...       0  2013-01-01 07:00:00  United Air Lines Inc.
18  2013      1    1  ...      13  2013-01-01 07:00:00  United Air Lines Inc.
19  2013      1    1  ...      25  2013-01-01 07:00:00  United Air Lines Inc.
20  2013      1    1  ...      30  2013-01-01 07:00:00  United Air Lines Inc.
21  2013      1    1  ...      32  2013-01-01 07:00:00  United Air Lines Inc.
22  2013      1    1  ...      45  2013-01-01 06:00:00  United Air Lines Inc.
23  2013      1    1  ...      39  2013-01-01 07:00:00  United Air Lines Inc.
24  2013      1    1  ...      46  2013-01-01 07:00:00  United Air Lines Inc.
25  2013      1    1  ...      50  2013-01-01 07:00:00  United Air Lines Inc.
26  2013      1    1  ...       0  2013-01-01 08:00:00  United Air Lines Inc.
27  2013      1    1  ...      23  2013-01-01 08:00:00  United Air Lines Inc.
28  2013      1    1  ...      17  2013-01-01 08:00:00  United Air Lines Inc.
29  2013      1    1  ...      23  2013-01-01 08:00:00  United Air Lines Inc.

[30 rows x 20 columns]

# Are there any missing values in this DataFrame? If so, drop them.

In [ ]:
nycflights13.isnull()

year  month    day  dep_time  ...   hour  minute  time_hour   name
0       False  False  False     False  ...  False   False      False  False
1       False  False  False     False  ...  False   False      False  False
2       False  False  False     False  ...  False   False      False  False
3       False  False  False     False  ...  False   False      False  False
4       False  False  False     False  ...  False   False      False  False
...       ...    ...    ...       ...  ...    ...     ...        ...    ...
336771  False  False  False     False  ...  False   False      False  False
336772  False  False  False     False  ...  False   False      False  False
336773  False  False  False     False  ...  False   False      False  False
336774  False  False  False     False  ...  False   False      False  False
336775  False  False  False     False  ...  False   False      False  False

[336776 rows x 20 columns]

At first glance it look like there are not any missing values, but this is a very large database. Just in case we will run the dropna() method



In [ ]:
nycflights13_filtered = nycflights13.dropna()
nycflights13_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 327346 entries, 0 to 336775
Data columns (total 20 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   year            327346 non-null  int64  
 1   month           327346 non-null  int64  
 2   day             327346 non-null  int64  
 3   dep_time        327346 non-null  float64
 4   sched_dep_time  327346 non-null  int64  
 5   dep_delay       327346 non-null  float64
 6   arr_time        327346 non-null  float64
 7   sched_arr_time  327346 non-null  int64  
 8   arr_delay       327346 non-null  float64
 9   carrier         327346 non-null  object 
 10  flight          327346 non-null  int64  
 11  tailnum         327346 non-null  object 
 12  origin          327346 non-null  object 
 13  dest            327346 non-null  object 
 14  air_time        327346 non-null  float64
 15  distance        327346 non-null  int64  
 16  hour            327346 non-null  int64  
 17  minute    

There were some missing values in the original database, using dropna() we have removed them. 

# **census**

# Merge these two DataFrames into a single DataFrame called census.

In [ ]:
population = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/state-populations.csv')
regions = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/census-divisions.csv')

census = pd.merge(population, regions)
census.info()
census.head(51)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 51 entries, 0 to 50
Data columns (total 10 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   state     51 non-null     object
 1   2010      51 non-null     int64 
 2   2011      51 non-null     int64 
 3   2012      51 non-null     int64 
 4   2013      51 non-null     int64 
 5   2014      51 non-null     int64 
 6   2015      51 non-null     int64 
 7   2016      51 non-null     int64 
 8   region    51 non-null     object
 9   division  51 non-null     object
dtypes: int64(7), object(3)
memory usage: 4.4+ KB


state      2010  ...     region            division
0                Alabama   4785492  ...      South  East South Central
1                 Alaska    714031  ...       West             Pacific
2                Arizona   6408312  ...       West            Mountain
3               Arkansas   2921995  ...      South  West South Central
4             California  37332685  ...       West             Pacific
5               Colorado   5048644  ...       West            Mountain
6            Connecticut   3579899  ...  Northeast         New England
7               Delaware    899816  ...      South      South Atlantic
8   District of Columbia    605183  ...      South      South Atlantic
9                Florida  18849098  ...      South      South Atlantic
10               Georgia   9713521  ...      South      South Atlantic
11                Hawaii   1363945  ...       West             Pacific
12                 Idaho   1571010  ...       West            Mountain
13              Illinois  12841578  ...    Midwest  East North Central
14               Indiana   6490528  ...    Midwest  East North Central
15                  Iowa   3050738  ...    Midwest  West North Central
16                Kansas   2858850  ...    Midwest  West North Central
17              Kentucky   4348662  ...      South  East South Central
18             Louisiana   4544996  ...      South  West South Central
19                 Maine   1327730  ...  Northeast         New England
20              Maryland   5788584  ...      South      South Atlantic
21         Massachusetts   6565524  ...  Northeast         New England
22              Michigan   9877495  ...    Midwest  East North Central
23             Minnesota   5311147  ...    Midwest  West North Central
24           Mississippi   2970322  ...      South  East South Central
25              Missouri   5996118  ...    Midwest  West North Central
26               Montana    990641  ...       West            Mountain
27              Nebraska   1830051  ...    Midwest  West North Central
28                Nevada   2703284  ...       West            Mountain
29         New Hampshire   1316872  ...  Northeast         New England
30            New Jersey   8803729  ...  Northeast     Middle Atlantic
31            New Mexico   2064756  ...       West            Mountain
32              New York  19402640  ...  Northeast     Middle Atlantic
33        North Carolina   9558915  ...      South      South Atlantic
34          North Dakota    674526  ...    Midwest  West North Central
35                  Ohio  11540983  ...    Midwest  East North Central
36              Oklahoma   3759603  ...      South  West South Central
37                Oregon   3838048  ...       West             Pacific
38          Pennsylvania  12712343  ...  Northeast     Middle Atlantic
39          Rhode Island   1053337  ...  Northeast         New England
40        South Carolina   4635943  ...      South      South Atlantic
41          South Dakota    816325  ...    Midwest  West North Central
42             Tennessee   6356671  ...      South  East South Central
43                 Texas  25244310  ...      South  West South Central
44                  Utah   2775326  ...       West            Mountain
45               Vermont    625982  ...  Northeast         New England
46              Virginia   8025773  ...      South      South Atlantic
47            Washington   6743226  ...       West             Pacific
48         West Virginia   1854230  ...      South      South Atlantic
49             Wisconsin   5690263  ...    Midwest  East North Central
50               Wyoming    564513  ...       West            Mountain

[51 rows x 10 columns]

# Reshape census so that one column contains all population measures, and another contains all of the year attributes.

In [ ]:
census_melted = pd.melt(frame=census, id_vars=['state', 'region', 'division'], value_vars=['2010', '2011', '2012', '2013', '2014', '2015', '2016'], var_name='year', value_name='population')
census_melted

state   region            division  year  population
0          Alabama    South  East South Central  2010     4785492
1           Alaska     West             Pacific  2010      714031
2          Arizona     West            Mountain  2010     6408312
3         Arkansas    South  West South Central  2010     2921995
4       California     West             Pacific  2010    37332685
..             ...      ...                 ...   ...         ...
352       Virginia    South      South Atlantic  2016     8411808
353     Washington     West             Pacific  2016     7288000
354  West Virginia    South      South Atlantic  2016     1831102
355      Wisconsin  Midwest  East North Central  2016     5778708
356        Wyoming     West            Mountain  2016      585501

[357 rows x 5 columns]

# Group the data by year and summarize it.

In [ ]:
census_melted.groupby(["year"]).describe()


population                              ...                                  
          count          mean           std  ...        50%        75%         max
year                                         ...                                  
2010       51.0  6.065651e+06  6.839536e+06  ...  4348662.0  6654375.0  37332685.0
2011       51.0  6.111046e+06  6.907730e+06  ...  4369354.0  6717221.5  37676861.0
2012       51.0  6.156831e+06  6.974942e+06  ...  4384799.0  6776617.0  38011074.0
2013       51.0  6.200096e+06  7.038774e+06  ...  4400477.0  6837396.0  38335203.0
2014       51.0  6.246342e+06  7.110811e+06  ...  4413057.0  6902053.5  38680810.0
2015       51.0  6.292091e+06  7.181603e+06  ...  4424611.0  6988927.5  38993940.0
2016       51.0  6.335834e+06  7.243849e+06  ...  4436974.0  7109535.5  39250017.0

[7 rows x 8 columns]

# Group the data by region, division, and year. Summarize it.

In [ ]:
pd.set_option('display.max_rows', None) 

census_grouped = census_melted.groupby(['region', 'division', 'year']).describe()
census_grouped


population  ...            
                                       count  ...         max
region    division           year             ...            
Midwest   East North Central 2010  5.0        ...  12841578.0
                             2011  5.0        ...  12860012.0
                             2012  5.0        ...  12870798.0
                             2013  5.0        ...  12879505.0
                             2014  5.0        ...  12867544.0
                             2015  5.0        ...  12839047.0
                             2016  5.0        ...  12801539.0
          West North Central 2010  7.0        ...  5996118.0 
                             2011  7.0        ...  6010717.0 
                             2012  7.0        ...  6025415.0 
                             2013  7.0        ...  6042711.0 
                             2014  7.0        ...  6060930.0 
                             2015  7.0        ...  6076204.0 
                             2016  7.0        ...  6093000.0 
Northeast Middle Atlantic    2010  3.0        ...  19402640.0
                             2011  3.0        ...  19519529.0
                             2012  3.0        ...  19602769.0
                             2013  3.0        ...  19673546.0
                             2014  3.0        ...  19718515.0
                             2015  3.0        ...  19747183.0
                             2016  3.0        ...  19745289.0
          New England        2010  6.0        ...  6565524.0 
                             2011  6.0        ...  6611923.0 
                             2012  6.0        ...  6658008.0 
                             2013  6.0        ...  6706786.0 
                             2014  6.0        ...  6749911.0 
                             2015  6.0        ...  6784240.0 
                             2016  6.0        ...  6811779.0 
South     East South Central 2010  4.0        ...  6356671.0 
                             2011  4.0        ...  6397634.0 
                             2012  4.0        ...  6454306.0 
                             2013  4.0        ...  6494821.0 
                             2014  4.0        ...  6544663.0 
                             2015  4.0        ...  6595056.0 
                             2016  4.0        ...  6651194.0 
          South Atlantic     2010  9.0        ...  18849098.0
                             2011  9.0        ...  19096952.0
                             2012  9.0        ...  19344156.0
                             2013  9.0        ...  19582022.0
                             2014  9.0        ...  19888741.0
                             2015  9.0        ...  20244914.0
                             2016  9.0        ...  20612439.0
          West South Central 2010  4.0        ...  25244310.0
                             2011  4.0        ...  25646389.0
                             2012  4.0        ...  26071655.0
                             2013  4.0        ...  26473525.0
                             2014  4.0        ...  26944751.0
                             2015  4.0        ...  27429639.0
                             2016  4.0        ...  27862596.0
West      Mountain           2010  8.0        ...  6408312.0 
                             2011  8.0        ...  6467163.0 
                             2012  8.0        ...  6549634.0 
                             2013  8.0        ...  6624617.0 
                             2014  8.0        ...  6719993.0 
                             2015  8.0        ...  6817565.0 
                             2016  8.0        ...  6931071.0 
          Pacific            2010  5.0        ...  37332685.0
                             2011  5.0        ...  37676861.0
                             2012  5.0        ...  38011074.0
                             2013  5.0        ...  38335203.0
                             2014  5.0        ...  38680810.0
                             2015  5.0        ...  38993940.0
    